servlet/src/main/java/com/linecorp/armeria/server/servlet/DefaultServletRegistration.java

```diff
@@ -0,0 +1,181 @@
+
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.Enumeration;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Set;
+
+import javax.annotation.Nullable;
+import javax.servlet.MultipartConfigElement;
+import javax.servlet.Servlet;
+import javax.servlet.ServletConfig;
+import javax.servlet.ServletContext;
+import javax.servlet.ServletRegistration.Dynamic;
+import javax.servlet.ServletSecurityElement;
+
+import com.google.common.collect.ImmutableMap;
+import com.google.common.collect.ImmutableSet;
+
+/**
+ * The servlet supportPipeline.
+ */
+final class DefaultServletRegistration implements Dynamic {
+    private final String servletName;
+    private final Servlet servlet;
+    private final ServletConfig servletConfig;
+    private final UrlMapper<DefaultServletRegistration> urlMapper;
+    private final Set<String> mappingSet = new HashSet<>();
+    private final Map<String, String> initParameterMap;
+
+    /**
+     * Creates a new instance.
+     */
+    DefaultServletRegistration(String servletName, Servlet servlet, DefaultServletContext servletContext,
+                               UrlMapper<DefaultServletRegistration> urlMapper,
+                               Map<String, String> initParameterMap) {
+        requireNonNull(servletName, "servletName");
+        requireNonNull(servlet, "servlet");
+        requireNonNull(servletContext, "servletContext");
+        requireNonNull(urlMapper, "urlMapper");
+
+        this.servletName = servletName;
+        this.servlet = servlet;
+        this.urlMapper = urlMapper;
+        this.initParameterMap = ImmutableMap.copyOf(initParameterMap);
+        servletConfig = new ServletConfig() {
+            @Override
+            public String getServletName() {
+                return servletName;
+            }
+
+            @Override
+            public ServletContext getServletContext() {
+                return servletContext;
+            }
+
+            @Override
+            @Nullable
+            public String getInitParameter(String name) {
+                return initParameterMap.get(name);
+            }
+
+            @Override
+            public Enumeration<String> getInitParameterNames() {
+                return Collections.enumeration(ImmutableSet.copyOf(getInitParameters().keySet()));
+            }
+        };
+        addMapping(servletName);
+    }
+
+    /**
+     * Get servlet config.
+     */
+    ServletConfig getServletConfig() {
+        return servletConfig;
+    }
+
+    /**
+     * Get servlet.
+     */
+    Servlet getServlet() {
+        return servlet;
+    }
+
+    @Override
+    public Set<String> addMapping(String... urlPatterns) {
+        requireNonNull(urlPatterns, "urlPatterns");
+        mappingSet.addAll(Arrays.asList(urlPatterns));
+        for (String pattern : urlPatterns) {
+            urlMapper.addMapping(pattern, this, servletName);
+        }
+        return mappingSet;
```

Comment:
> yes, each servlet have one ServletRegistration to addMapping multiple urlPatterns to one servletName such as: urlPattern ("/home", "/home/user", "/home/*") map to servletName ("home").
I think, mappingSet is returned to check which urlPatterns were mapped.

So this shouldn't be:
```java
requireNonNull(urlPatterns, "urlPatterns");
final Set<String> current = ImmutableSet.copyOf(mappingSet);
mappingSet.addAll(Arrays.asList(urlPatterns));
for (String pattern : urlPatterns) {
    urlMapper.addMapping(pattern, this, servletName);
}
return current;
```
?

Reply:
I think, source code's logic is right. Because `mappingSet` only store `urlPatterns` of one `servletRegistration` and `urlMapper` store `urlPatterns` of all `servletRegistration`. I think we only need to add return if ServletContext is initialized at here.

Reply:
Do you mean we should return current (mappingSet before add urlPatterns) instead of (mappingSet after add urlPattern)?

Reply:
Ah, I misunderstand the Javadoc. Then, I'm wondering if we follow what Tomcat does: https://github.com/quickhack/tomcat/blob/master/src/main/java/org/apache/catalina/core/ApplicationServletRegistration.java#L166
🤔